lambda/matching.ml

```diff
@@ -2752,57 +2752,42 @@ let mk_failaction_neg partial ctx def =
     )
   | Total -> (None, Jumps.empty)
 
-(* In line with the article and simpler than before *)
+(* In [mk_failaction_pos partial seen ctx defs],
+   - [partial] is Total if we know the current switch
+     is already exhaustive
+   - [seen] is the list of constructors accepted by the switch
+     (those that will be matched)
+   - [ctx] is the current context (what we know of the value
+     being matched)
+   - [defs] is the default environment (what inputs
+     are expected by the switches present at larger exit numbers).
+
+   The function returns a triple [(fail, fails, jumps)] containing
+   information for the failure cases, the constructors missing from
+   the current switch:
+   - [fail] is an optional 'default' action for the switch
+   - [fails] is a list of extra switch clauses to add for failure cases
+   - [jumps] contains a jump summary for all these new cases
+     (context information for all exits they reach)
+
+   The general strategy is to compute an accurate list of [fails] and
+   try to avoid having a default action, as this generates better
+   code. But we choose to have a default action when the list [fails]
+   would be too large or too costly to compute.
+
+   Through its jump summary, [mk_failaction_pos] propagates "negative
+   information" about the constructors not taken. For example, if
+   a first submatrix contains a match on the [None] constructor, the
+   exit point selected for failure will get provided the context
+   information that its value must be [Some _].
```

Comment:
The last sentence isn't quite accurate. Jumps from other handlers to that exit point may not be `Some _`. It's more that jumps *from this switch* to that exit point must be `Some _`.

Reply:
Attempted reformulation:

> For example, if a switch only accepts the [None] constructor, [mk_failaction_pos] generates a failure clause along with context information that the value reaching the failure clause must be [Some _].